internal/cmd/kubernetes.go

```diff
@@ -85,80 +85,78 @@ func writeKubeconfig(user *api.User, destinations []api.Destination, grants []ap
 		return err
 	}
 
-	keep := make(map[string]bool)
+	type clusterContext struct {
+		Namespace string
+		URL string
+		CA []byte
+	}
+
+	infraContexts := make(map[string]clusterContext)
 
 	for _, g := range grants {
 		parts := strings.Split(g.Resource, ".")
-
 		cluster := parts[0]
 
 		var namespace string
 		if len(parts) > 1 {
 			namespace = parts[1]
 		}
 
-		context := "infra:" + cluster
-
-		if namespace != "" {
-			context += ":" + namespace
-		}
-
-		var (
-			url    string
-			ca     []byte
-			exists bool
-		)
-
+		var infraContext clusterContext
 		for _, d := range destinations {
 			if !isResourceForDestination(g.Resource, d.Name) {
 				continue
 			}
 
 			if isDestinationAvailable(d) {
-				url = d.Connection.URL
-				ca = []byte(d.Connection.CA)
-				exists = true
+				infraContext = clusterContext{
+					URL:  d.Connection.URL,
+					CA: []byte(d.Connection.CA),
+				}
 				break
 			}
 		}
 
-		if !exists {
+		if infraContext.URL == "" {
 			continue
 		}
 
-		u, err := urlx.Parse(url)
+		contextName := "infra:" + cluster
+		if _, ok := infraContexts[contextName]; ok && namespace != "" {
```

Comment:
If there's a non-namespace grant, set that as the default context. Otherwise, pick a namespace